### PR TITLE
Jasmin code gen

### PR DESCRIPTION
**Pull request by ScumbagDog**
Har lavet en codegenerator til jasmin

### CODE REVIEW

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+import ast.type.DoubleLiteral;
+import ast.type.IdNode;
+import ast.type.IntegerLiteral;
+import ast.type.StringLiteral;
+
+public class JasminCodeGeneratorVisitor extends AstVisitor{
+
+	private PrintStream out;
+	private StringBuilder sb;
+	private StringBuilder functionStringBuilder; //needed to remember the functions and declare them after main
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Remove comment

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+		currentVariableEnvironment = new ArrayList<String>();
+		currentLocalFunctions = new ArrayList<String>();
+		sb = new StringBuilder();
+		functionStringBuilder = new StringBuilder();
+		this.programName = programName;
+		
+	}
+	@Override
+    public void visit(StartNode node) {
+		generatePrefixSetupCode();
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Fix indention

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+        node.getDcls().accept(this);
+        node.getStmts().accept(this);
+        generatePostfixSetupCode();
+        out.print(sb);
+    }
+
+    @Override
+    public void visit(BlockNode node) {
+    	// Need to remember the state of the environment, as we cannot rollback the changes made in this block otherwise
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Remove comment

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+        if (node.getReturnstmtNode() != null) {
+            node.getReturnstmtNode().accept(this);
+        }
+        currentVariableEnvironment = oldVariableEnvironment;
+        currentLocalFunctions = oldLocalFunctions;
+    }
+
+    @Override
+    public void visit(DclNode node) {
+    		currentVariableEnvironment.add(node.getID());
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Fix indention

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+        }
+    }
+
+    @Override
+    public void visit(While_stmtNode node) {
+    	int beginLabel = getNextLabel();
+    	int enterLabel = getNextLabel();
+    	appendLine("goto " + enterLabel);
+    	append(beginLabel + ": ");
+        node.getBlockNode().accept(this);
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Fix indention

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+    @Override
+    public void visit(LogicalAndExprNode node) {	
+        node.getLeftNode().accept(this);
+        generateCodeForLogicalNegation();
+        node.getRightNode().accept(this);
+        generateCodeForLogicalNegation();
+        generateCodeForLogicalOr();
+        generateCodeForLogicalNegation();
+        //my group: You can't just use formal logic to rewrite code you dislike
+        //DeMorgan's Laws: That's where you're wrong, kiddo
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Remove jokes

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+	private void decrementStack() {
+		currentStackSize--;
+		if (currentStackSize < 0) {
+			currentStackSize = 0;
+		}
+	}
+	private void appendComparisonBasedOnType(Type comparedType) {
+		switch(comparedType) {
+		case INT:
+        	appendLine("lcmp");
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Fix idention

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+            	append("I");
+            	break;
+            case DOUBLE:
+            	append("D");
+            	break;
+            case STRING:
+            	append("Ljava/lang/String;");
+            	break;
+            }
+            ;
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Remove unused `;`

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
+		appendLine(".super java/lang/Object");
+		appendLine(".method public static main([Ljava/lang/String;)V");
+	}
+	private void generatePostfixSetupCode() {
+        appendLine("return");
+        appendLine(".limit stack " + maxStackSize);
+        appendLine(".limit locals " + maxLocals);
+        appendLine(".end method");
+        sb.append(functionStringBuilder);
+	}
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Maybe add function which takes a `Type` and returns JVM Type? 
When doing that, you can just concat strings together to create the JVM instruction

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java [hunk 10]
```diff
@@ -0,0 +1,684 @@
...
+	private void incrementStack() {
+		currentStackSize++;
+		if(currentStackSize > maxStackSize) {
+			maxStackSize = currentStackSize;
+		}
+	}
+	private void decrementStack() {
+		currentStackSize--;
+		if (currentStackSize < 0) {
```
Comment (joandrsn) [inline src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: Should this happen? Why not just let it blow up in our tests?

File: src/main/java/astvisitors/JasminCodeGeneratorVisitor.java (hunk 6 above)
Comment (ScumbagDog) [reply src/main/java/astvisitors/JasminCodeGeneratorVisitor.java]: I refuse

**Pull request merged**